Haptics/app/src/main/java/com/example/android/haptics/samples/ui/expand/ExpandRoute.kt

```diff
@@ -0,0 +1,337 @@
+import com.example.android.haptics.samples.ui.theme.secondaryText
+
+/**
+ * The two possible states states of the expand shape.
+ */
+private enum class ExpandShapeState {
+    Collapsed,
+    Expanded
+}
+
+private const val DEFAULT_ANIMATE_TO_EXPANDED_DURATION_MS = 650
+private const val DEFAULT_ANIMATE_TO_COLLAPSED_DURATION_MS = 500
+// Add a tick after the animation between states has occurred and the controls/indicators appear.
+private const val ANIMATION_COMPLETE_TICK_DELAY_MS = 30
+
+// We animate size and thickness of the shape. As it gets larger we make the circular shape
+// thinner.
+private val EXPAND_SHAPE_COLLAPSED_SIZE = 64.dp
+private val EXPAND_SHAPE_COLLAPSED_THICKNESS = EXPAND_SHAPE_COLLAPSED_SIZE / 2
+
+private val EXPAND_SHAPE_EXPANDED_SIZE = EXPAND_SHAPE_COLLAPSED_SIZE * 5
+private val EXPAND_SHAPE_EXPANDED_THICKNESS = EXPAND_SHAPE_COLLAPSED_SIZE / 4
+
+// Representation of the primitive composition to be played by the vibrator when the indicator is
+// expanding.
+private val VIBRATION_DATA_FOR_EXPANDING = PrimitiveComposition(
+    arrayOf(
+        PrimitiveEffect(VibrationEffect.Composition.PRIMITIVE_SLOW_RISE, 0.3f),
+        PrimitiveEffect(VibrationEffect.Composition.PRIMITIVE_QUICK_FALL, 0.3f),
+        PrimitiveEffect(VibrationEffect.Composition.PRIMITIVE_TICK, 0.6f, ANIMATION_COMPLETE_TICK_DELAY_MS)
+    )
+)
+
+// Representation of the primitive composition to be played by the vibrator when the indicator is
+// collapsing.
+private val VIBRATION_DATA_FOR_COLLAPSING = PrimitiveComposition(
+    arrayOf(
+        PrimitiveEffect(VibrationEffect.Composition.PRIMITIVE_SLOW_RISE),
+        PrimitiveEffect(VibrationEffect.Composition.PRIMITIVE_TICK, 1f, ANIMATION_COMPLETE_TICK_DELAY_MS)
+    )
+)
+
+@Composable
+fun ExpandRoute(viewModel: ExpandViewModel) {
+    ExpandExampleScreen(messageToUser = viewModel.messageToUser)
+}
+
+@Composable
+fun ExpandExampleScreen(messageToUser: String) {
+    val vibrator = LocalContext.current.getSystemService(Context.VIBRATOR_SERVICE) as Vibrator
+
+    var currentState by remember { mutableStateOf(ExpandShapeState.Collapsed) }
+
+    var animateToExpandedDuration = DEFAULT_ANIMATE_TO_EXPANDED_DURATION_MS
+    var animateToCollapsedDuration = DEFAULT_ANIMATE_TO_COLLAPSED_DURATION_MS
+    // Since this is a demo app, we still show this view, even if all primitives aren't supported
+    // and the example's vibration won't work. In that case we still show the animation using
+    // defaults above.
+    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S &&
+        VIBRATION_DATA_FOR_EXPANDING.allPrimitivesSupported(vibrator) &&
+        VIBRATION_DATA_FOR_COLLAPSING.allPrimitivesSupported(vibrator)
+    ) {
+        animateToExpandedDuration = remember {
+            VIBRATION_DATA_FOR_EXPANDING.getDuration(vibrator)
+        }
+        animateToCollapsedDuration = remember {
+            VIBRATION_DATA_FOR_COLLAPSING.getDuration(vibrator)
+        }
```

Comment:
Please consider passing on the default values to the getDuration methods, so you don't need to do API and primitive checks here, you can delegate to the PrimitiveComposition object.

Reply:
good call